cocos/2d/utils/text-utils.ts

```diff
@@ -214,6 +214,192 @@ export function safeMeasureText (ctx: CanvasRenderingContext2D, string: string,
     return width;
 }
 
+export function getSymbolLength (str: string): number {
+    const length = str.length;
+    let len = 0;
+    let count = 0;
+    let start = 0;
+    let charCode = 0;
+    for (let i = 0; i < length; i++) {
+        charCode = str.charCodeAt(i);
+        if (charCode === 0x200d) {
+            len++;
+            continue;
+        }
+        if (charCode >= 0xd800 && charCode <= 0xdbff) {
+            len++;
+            charCode = str.charCodeAt(i + 1);
+            if (charCode >= 0xdc00 && charCode <= 0xdfff) {
+                if (i + 2 < length && str.charCodeAt(i + 2) === 0x200d) {
+                    len++;
+                } else {
+                    len++;
+                    start += len;
+                    count++;
+                    len = 0;
+                }
+                i++;
+                continue;
+            }
+        }
+        start = i + 1;
+        count++;
+        len = 0;
+    }
+    return count;
+}
+
+export function getSymbolAt (str: string, index: number): string  {
+    const length = str.length;
+    let len = 0;
+    let count = 0;
+    let start = 0;
+    let charCode = 0;
+    for (let i = 0; i < length; i++) {
+        charCode = str.charCodeAt(i);
+        if (charCode === 0x200d) {
+            len++;
+            continue;
+        }
+        if (charCode >= 0xd800 && charCode <= 0xdbff) {
+            len++;
+            charCode = str.charCodeAt(i + 1);
+            if (charCode >= 0xdc00 && charCode <= 0xdfff) {
+                if (i + 2 < length && str.charCodeAt(i + 2) === 0x200d) {
+                    len++;
+                } else {
+                    len++;
+                    if (index === count) {
+                        return str.slice(start, start + len);
+                    }
+                    start += len;
+                    count++;
+                    len = 0;
+                }
+                i++;
+                continue;
+            }
+        }
+        if (index === count) {
+            return str.charAt(i);
+        }
+        start = i + 1;
+        count++;
+        len = 0;
+    }
+    return '';
+}
+
+export function getSymbolCodeAt (str: string, index: number): string  {
```

Comment:
‘getSymbolAt’ and ‘getSymbolCodeAt’ have many similar codes. Can they share some codes?